### DISCUSSION TITLE
Questions regarding freeing resources reserved by MarshalInterface.FromManaged()

### DISCUSSION THREAD

**Discussion by mqudsi** [microsoft/CsWinRT#1191] [Q&A] [answered]
Hello team,

I'm making the transition from the old Framework-provided CCW marshalling to the new CsWinRt/WASDK-provided alternative and there's something I'm not clear about.

The old `Marshal.GetComInterfaceForObject()` approach [was well-documented](https://docs.microsoft.com/en-us/dotnet/api/system.runtime.interopservices.marshal.getcominterfaceforobject?view=net-6.0) and made it clear that pinned memory and resources for the CCW  are reserved until a paired call to `Marshal.Release()` was made to decrement the internal ref count and trigger the release of associated resources.

I can't for the life of me find any documentation on the new `MarshalInterface<T>` and co. but just going by the code, `MarshalInterface<TInterface>.FromManaged(new T())` returns a bare pointer (well, `IntPtr`) but it seems that lifetime tracking is provided by the `ObjectReference<TInterface>` class; all of which would imply that a bare `IntPtr` returned from `MarshalInterface.FromManaged()` doesn't have any sort of reference counting, which leads to some questions:

I'm assuming the IntPtr returned from `MarshalInterface.FromManaged()` is obviously pinned, or else it would be unsafe to directly use/return as a CCW. How is that memory unpinned and the CCW destroyed? Is that the job of `MarshalInterface.DisposeAbi()` (which seems like it's the counterpart to `MarshalInterface.FromAbi()` and not `MarshalInterface.FromManaged()`)?

[The interop docs](https://github.com/microsoft/CsWinRT/blob/master/docs/interop.md) say the new replacement for `Marshal.Release()` is only via the `ObjectInterface` class - does that mean that we actually can't use the `IntPtr` returned by `MI.FromManaged()` without first using `ObjectReference.Attach()` and then using `IObjectReference.GetRef()` to get the CCW? I'm guessing no because it would've made more sense to have `MI.FromManaged()` return an `IObjectReference` directly instead of an `IntPtr` in that case?

Basically, what do I have to do after using `MarshalInterface<TInterface>.FromManaged(new T())` in my `IClassFactory`/COM server to make sure nothing is leaked?

**Comment by manodasanW** [comment, marked as answer]
@mqudsi When you call `FromManaged` on a C# implemented object, a CCW is created for it (if none exists) for which the pointer is returned and the reference count on the CCW is incremented.  This pointer which is returned is basically the ABI representation of the object and also owns the reference count.  If you want to release the reference, you can make use of `DisposeAbi` as you mentioned or if you passed that pointer to the native side, the native side can call `Release` on the `IUnknown` interface just like any other COM object.
